Handle PermissionError when looking for the defaults or groups files

nahun opened the pull request:
When checking if the `defaults` or `groups` files exist, it can throw a permissions error. This can happen when running a script with sudo as a different user that doesn't have permission to the current working directory.

wvandeun commented:
Thanks for the suggestion.

I see your point but I do have some thoughts about the approach.
I'm not convinced at this stage that the inventory plugin should handle the PermissionError exception.
Wouldn't it be better to signal that exception to the consumer of the plugin that this error occurred? 

With your approach we are completely ignoring the error and not communicating the condition back to the consumer.
Continuing setting up the inventory would not be what I would expect.

The PermissionError could better be handled in the consuming script so it has control over what should happen (ie print/log an error message and exit).

Maybe I'm overseeing something or I may not have enough context to understand what you are trying to achieve?

nahun commented:
If it was required that the `groups` or `defaults` file existed and be readable then yes I think the error should be passed to the consuming script. But my thought was that it treats the PermissionError the same as if the file didn't exist. To the plugin, all that matters is that the file is readable (which means it exists and permissions are correct). I think a warning in the log is probably a good idea.

Basically I need the plugin to continue as normal because the script will very often be run in a current working directory that is not readable by the user (since its run with sudo). For example a user logs into our server, gets placed into their home directory, and then just runs the script from there with something like `sudo -u otheruser thescript`. The "otheruser" is not root and does not have permissions to the user's home directory.

If I'm able to make the plugin ignore the PermissionsError another way (via an option) that would be fine as well.

wvandeun commented:
I would be open to this, if the default behaviour of the plugin doesn't change. The option should be disabled by default. A logging message should be created (warning level?) when the option gets enabled.
Also tests would need to be added for this behaviour.

nahun commented:
I updated the branch. Something like that?

wvandeun commented on nornir_netbox/plugins/inventory/netbox.py:
```diff
@@ -308,22 +310,32 @@ def load(self) -> Inventory:
+                defaults = _get_defaults(defaults_dict)
+        except PermissionError as e:
+            if not self.ignore_file_errors:
+                raise PermissionError(e)
```
`raise` should be sufficient here to re-raise the current handled exception
would add a logging message (error level)

wvandeun commented on nornir_netbox/plugins/inventory/netbox.py:
```diff
@@ -308,22 +310,32 @@ def load(self) -> Inventory:
-        if self.group_file.exists():
-            with self.group_file.open("r") as f:
-                groups_dict = yml.load(f) or {}
+            warnings.warn("Error ignored while reading file: {}".format(e))
```
Can we use the logging library here instead?
The message can be more explicit, something like `Unable to read {file} due to a permission issue`

wvandeun commented on nornir_netbox/plugins/inventory/netbox.py:
```diff
@@ -308,22 +310,32 @@ def load(self) -> Inventory:
+                    g.groups = ParentGroups([groups[g] for g in g.groups])
+        except PermissionError as e:
+            if not self.ignore_file_errors:
+                raise PermissionError(e)
```
same remark as above

wvandeun reviewed (changes requested):
Can you have a look at the comments and also document the newly added option?
Still need to review the tests.

nahun commented:
Pushed changes that should cover those comments. Thanks!

wvandeun commented:
Made a few suggestions/improvements.
Are you ok with these?

nahun commented:
Looks good, didn't know about the `tmp_path` fixture with pytest. Thanks!

wvandeun committed:
adds ignore_file_permission_errors inventory option

When enabling this option the inventory will continue loading if there
is a permission error occurring when reading the group or defaults file.

wvandeun committed:
adds docs for ignore_file_permission_errors option

wvandeun committed:
adds tests for ignore_file_permission_errors option

wvandeun commented:
Thank you for your contribution on this one!
